### PR TITLE
Implemented functionality to skip the cache for MI when claims are provided

### LINKED ISSUE #7206: [Engineering task] Implement Claims API to Bypass Cache When Claims are Present in MSAL with Managed Identity
### Core Library

MSAL Node (@azure/msal-node)

### Wrapper Library

MSAL Node Extensions (@azure/msal-node-extensions)

### Public or Confidential Client?

Confidential

### MSAL client type

Managed identity

### MSAL client type

Confidential

### Problem Statement

**Task type**
Development

**Description**
Currently, MSAL with Managed Identity does not expose any API claims API. With CAE (Continuous Access Evaluation) being enabled by default, we need to implement a mechanism to bypass the cache if claims are detected in the token request.

**Steps to Reproduce:**
- Enable CAE by default in MSAL with Managed Identity.
- Make a token request with claims present.
- By-pass cache when claims are present

note : msi v1 endpoint is unchanged so there is no need to pass any claims to the endpoint itself, this feature is done so MSAL will bypass the cache. 

Observe that the cache is not bypassed, leading to potential stale token usage.

**Expected Behavior:**
When claims are present in the token request, the cache should be bypassed to ensure that the latest token is used, in line with CAE requirements.

### Proposed solution

- Expose the claims API in MSAL for MI
- Expose Claims to MI Assertion Provider for FIC

### Alternatives

_No response_

### Source

Internal (Microsoft)

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**
Re-used functionality from ClientCredential flow.

This PR originally contained code to deprecate client assertion strings. That will now be a separate PR.

### CODE REVIEW

File: lib/msal-node/apiReview/msal-node.api.md [hunk 1]
```diff
@@ -449,6 +449,7 @@ export type ManagedIdentityIdParams = {
 //
 // @public
 export type ManagedIdentityRequestParams = {
+    claims?: string;
```
Comment (neha-bhargava) [inline lib/msal-node/apiReview/msal-node.api.md:452]: My understanding is that the claims also needs to be passed to the managed identity endpoint. @gladjohn can you confirm that?

Comment (Robbie-Microsoft) [reply lib/msal-node/apiReview/msal-node.api.md:452]: oops. Yea, I'll have to add them to the endpoint.

Comment (Robbie-Microsoft) [reply lib/msal-node/apiReview/msal-node.api.md:452]: done

File: lib/msal-common/apiReview/msal-common.api.md [hunk 2]
```diff
@@ -3461,6 +3461,23 @@ export type RequestThumbprint = {
     shrOptions?: ShrOptions;
 };
 
+// Warning: (ae-missing-release-tag) "RequestValidator" is part of the package's API, but it is missing a release tag (@alpha, @beta, @public, or @internal)
+//
+// @public
+export class RequestValidator {
```
Comment (bgavrilMS) [inline lib/msal-common/apiReview/msal-common.api.md]: Why is this class public?

File: change/@azure-msal-common-e4f59bc0-0846-4b3a-a6bd-b8a4829c5bff.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Implemented functionality to skip the cache for MI when claims are provided #7207",
```
Comment (bgavrilMS) [inline change/@azure-msal-common-e4f59bc0-0846-4b3a-a6bd-b8a4829c5bff.json]: This is a good first PR, but it does not fully resolve the issue. The claims must be injected in leg1 and leg2 of FIC - this is separate work.

**Comment by bgavrilMS** [comment]
@Robbie-Microsoft - pls continue to the work we discussed, but do not commit. This work hasn't been OK-ed yet by the MSI team. It will anyway be required for MSIv2.

@gladjohn  - please close the work items until we get confirmation.

File: lib/msal-common/apiReview/msal-common.api.md (hunk 2 above)
Comment (Robbie-Microsoft) [reply lib/msal-common/apiReview/msal-common.api.md:3468]: This change was auto-generated. The only change I made was export it in msal-common so it can be used in msal-node.

**Comment by gladjohn** [comment]
> @gladjohn - please close the work items until we get confirmation.

MSI team has put this work on hold for now, no further details yet.

File: lib/msal-node/apiReview/msal-node.api.md (hunk 1 above)
Comment (gladjohn) [reply lib/msal-node/apiReview/msal-node.api.md:452]: MI v1 endpoint do not take any claims, the ask is to simply bypass cache with claims is passed to MSAL.

File: lib/msal-common/apiReview/msal-common.api.md (hunk 2 above)
Comment (bgavrilMS) [reply lib/msal-common/apiReview/msal-common.api.md:3471]: I think this needs to be removed?

File: lib/msal-common/src/account/ClientCredentials.ts [hunk 4]
```diff
@@ -6,17 +6,23 @@
...
+/**
+ * @deprecated Provide a callback instead of a string
+ */
+type assertionString = string;
+type assertionCallback = ClientAssertionCallback;
 /**
  * Client Assertion credential for Confidential Clients
  */
 export type ClientAssertion = {
-    assertion: string | ClientAssertionCallback;
```
Comment (bgavrilMS) [inline lib/msal-common/src/account/ClientCredentials.ts]: I don't think you can change the type names ? Isn't it a breaking change? You can easily check by having some test code that checks the old version and compile against this new version.

File: lib/msal-common/src/index.ts [hunk 5]
```diff
@@ -140,6 +140,7 @@ export { NativeRequest } from "./request/NativeRequest";
 export { NativeSignOutRequest } from "./request/NativeSignOutRequest";
 export { RequestParameterBuilder } from "./request/RequestParameterBuilder";
 export { StoreInCache } from "./request/StoreInCache";
+export { RequestValidator } from "./request/RequestValidator";
```
Comment (bgavrilMS) [inline lib/msal-common/src/index.ts]: Why was this added?

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts [hunk 6]
```diff
@@ -138,6 +140,12 @@ export abstract class BaseManagedIdentitySource {
 
         const networkRequestOptions: NetworkRequestOptions = { headers };
 
+        if (managedIdentityRequest.claims) {
+            RequestValidator.validateClaims(managedIdentityRequest.claims);
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: What does this do? Afaik claims is just a string (a json string) from MSAL's point of view. Does this need validaton?

File: lib/msal-common/apiReview/msal-common.api.md [hunk 7]
```diff
@@ -4208,6 +4226,8 @@ const X_MS_LIB_CAPABILITY = "x-ms-lib-capability";
 
 // Warnings were encountered during analysis:
 //
+// src/account/ClientCredentials.ts:25:5 - (ae-forgotten-export) The symbol "assertionString" needs to be exported by the entry point index.d.ts
```
Comment (bgavrilMS) [inline lib/msal-common/apiReview/msal-common.api.md]: This needs to be resolved.

File: lib/msal-common/apiReview/msal-common.api.md [hunk 8]
```diff
@@ -4208,6 +4226,8 @@ const X_MS_LIB_CAPABILITY = "x-ms-lib-capability";
...
+// src/account/ClientCredentials.ts:25:5 - (ae-forgotten-export) The symbol "assertionCallback" needs to be exported by the entry point index.d.ts
```
Comment (bgavrilMS) [inline lib/msal-common/apiReview/msal-common.api.md]: This needs to be resolved.

File: lib/msal-common/apiReview/msal-common.api.md [hunk 9]
```diff
@@ -1057,7 +1057,7 @@ const CLIENT_SECRET = "client_secret";
 //
```
Comment (bgavrilMS) [inline lib/msal-common/apiReview/msal-common.api.md]: Pls take some time to resolve all these warnings.

File: lib/msal-common/src/account/ClientCredentials.ts [hunk 10]
```diff
@@ -6,17 +6,23 @@
...
     tokenEndpoint?: string;
+    claims?: string;
 };
 
 export type ClientAssertionCallback = (
     config: ClientAssertionConfig
 ) => Promise<string>;
 
+/**
+ * @deprecated Provide a callback instead of a string
```
Comment (bgavrilMS) [inline lib/msal-common/src/account/ClientCredentials.ts]: I assume you did this because you can't just deprecate "string"?

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
I am not sure if this doesn't introduce a breaking change.

File: lib/msal-common/src/account/ClientCredentials.ts (hunk 4 above)
Comment (Robbie-Microsoft) [reply lib/msal-common/src/account/ClientCredentials.ts:19]: I'm not sure it's a breaking change, but it's a way I can deprecate the string assertion. 

Maybe alternatively, I can leave "assertion" as it is and deprecate it, then create a new property "assertionCallback" that is ONLY type ClientAssertionCallback.

Thoughts?

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts (hunk 6 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: This method makes sure the claims string provided by a developer is a JSON string. It ensures the claims string can be parsed via `JSON.parse(claimsString)` and throws an error if it can't be parsed.

File: lib/msal-common/apiReview/msal-common.api.md (hunk 2 above)
Comment (Robbie-Microsoft) [reply lib/msal-common/apiReview/msal-common.api.md]: It's exported in msal-common's index.ts. That makes it public.

File: lib/msal-common/src/index.ts (hunk 5 above)
Comment (Robbie-Microsoft) [reply lib/msal-common/src/index.ts]: We're using it in msal-node now to validate the claims string provided by a developer is a JSON string

File: lib/msal-common/src/account/ClientCredentials.ts (hunk 4 above)
Comment (Robbie-Microsoft) [reply lib/msal-common/src/account/ClientCredentials.ts]: I'll be making a new PR for this deprecation.

File: lib/msal-common/apiReview/msal-common.api.md [hunk 11]
```diff
@@ -4147,7 +4157,7 @@ export type ValidCredentialType = IdTokenEntity | AccessTokenEntity | RefreshTok
 // Warning: (ae-missing-release-tag) "version" is part of the package's API, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export const version = "14.14.0";
+export const version = "14.14.1";
```
Comment (Robbie-Microsoft) [inline lib/msal-common/apiReview/msal-common.api.md]: This change is due to a manual release last week. Hector is going to make a PR to bump the versions.

File: lib/msal-node/apiReview/msal-node.api.md [hunk 12]
```diff
@@ -732,7 +733,7 @@ export { ValidCacheType }
 // Warning: (ae-missing-release-tag) "version" is part of the package's API, but it is missing a release tag (@alpha, @beta, @public, or @internal)
 //
 // @public (undocumented)
-export const version = "2.12.0";
+export const version = "2.13.0";
```
Comment (Robbie-Microsoft) [inline lib/msal-node/apiReview/msal-node.api.md]: This change is due to a manual release last week. Hector is going to make a PR to bump the versions.

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts (hunk 6 above)
Comment (bgavrilMS) [reply lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: This validation adds little value and increases client latency. ESTS can reject the claims. Pls remove.

File: lib/msal-common/apiReview/msal-common.api.md (hunk 2 above)
Comment (bgavrilMS) [reply lib/msal-common/apiReview/msal-common.api.md]: I don't think this is good enough reason. But in any case, there is no need to validate claims.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
don't validate claims

**Review by bgavrilMS — APPROVED** [review]

File: lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts [hunk 13]
```diff
@@ -138,6 +139,11 @@ export abstract class BaseManagedIdentitySource {
 
         const networkRequestOptions: NetworkRequestOptions = { headers };
 
+        if (managedIdentityRequest.claims) {
+            networkRequest.bodyParameters[AADServerParamKeys.CLAIMS] =
```
Comment (neha-bhargava) [inline lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: Not sure if this is right. Are we sending the claims to the endpoint? Maybe I am missing something.

Comment (gladjohn) [reply lib/msal-node/src/client/ManagedIdentitySources/BaseManagedIdentitySource.ts]: this design has been changing, and is still not final. Initially the ask was just to bypass cache and get a new token from the IMDS endpoint when claims are passed. So this is good as is. But there is a new change coming and more details to follow, but for now let's just skip the cache and get a new token.

File: lib/msal-node/src/client/ManagedIdentityApplication.ts [hunk 14]
```diff
@@ -142,7 +142,10 @@ export class ManagedIdentityApplication {
             correlationId: this.cryptoProvider.createNewGuid(),
         };
 
-        if (managedIdentityRequest.forceRefresh) {
+        if (
+            managedIdentityRequestParams.claims ||
```
Comment (gladjohn) [inline lib/msal-node/src/client/ManagedIdentityApplication.ts:146]: does msal js already have a way for app developers to provide this ?

Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ManagedIdentityApplication.ts:146]: yes, but only per request basis. Hence, me adding the claims to the MI request params.

File: lib/msal-node/src/request/ManagedIdentityRequestParams.ts [hunk 15]
```diff
@@ -5,10 +5,12 @@
 
 /**
  * ManagedIdentityRequest
+ * - claims       - a stringified claims request which will be used to determine whether or not the cache should be skipped
```
Comment (bgavrilMS) [inline lib/msal-node/src/request/ManagedIdentityRequestParams.ts:8]:
```suggestion
 * - claims       - in case a resource refuses to authenticate with a token (e.g. because the token has been revoked by the tenant admin), it may send a 401 response and a WWW-Authenticate header with a claims parameter. When this is configured, MSAL always requests a new token from the identity provider, and will add the claims to the request where this is supported.
```

**Review by bgavrilMS — APPROVED** [review]

**Review by gladjohn — APPROVED** [review]
